File: apps/android/feature/ai/data/src/main/java/com/scribblefit/feature/ai/data/engine/ScribbleFitProxyEngine.kt
package com.scribblefit.feature.ai.data.engine

import com.scribblefit.core.network.ScribbleFitApi
import com.scribblefit.core.network.model.ParseRequest
import com.scribblefit.feature.ai.data.mapper.toDomain
import com.scribblefit.feature.ai.domain.engine.AnalysisEngine
import com.scribblefit.feature.ai.domain.engine.ConfigRepository
import com.scribblefit.feature.ai.domain.engine.LLMEngine
import com.scribblefit.feature.ai.domain.model.AnalysisSuggestion
import com.scribblefit.feature.ai.domain.model.AnalysisSummary
import com.scribblefit.feature.ai.domain.model.ExerciseInsight
import com.scribblefit.feature.ai.domain.model.ParsedWorkoutResult
import com.scribblefit.feature.ai.domain.model.ParsingStatus
import com.scribblefit.feature.ai.domain.model.SummaryPeriod
import com.scribblefit.feature.ai.domain.security.SecureKeyStorage
import kotlinx.coroutines.flow.first
import org.slf4j.LoggerFactory
import javax.inject.Inject

class ScribbleFitProxyEngine @Inject constructor(
    private val api: ScribbleFitApi,
    private val secureKeyStorage: SecureKeyStorage,
    private val configRepository: ConfigRepository
) : LLMEngine, AnalysisEngine {

    private val logger = LoggerFactory.getLogger(javaClass)

    override suspend fun parseWorkout(rawText: String): ParsedWorkoutResult {
        val startTime = System.currentTimeMillis()
        return try {
            val token = secureKeyStorage.getAuthToken()
            val systemPrompt = configRepository.getConfig().first()?.promptText
                ?: error("Prompt is empty. Configuration is not set.")
            val request = ParseRequest(rawText = rawText, prompt = systemPrompt)

            val parsedWorkoutDto = api.parseProxy(request, token)
            val workout = parsedWorkoutDto.toDomain()
            val duration = System.currentTimeMillis() - startTime

            ParsedWorkoutResult(
                workout = workout,
                rawText = rawText,
                status = ParsingStatus.SUCCESS,
                modelUsed = "proxy-orchestrator",
                processingTimeMs = duration
            )
        } catch (e: Exception) {
            val duration = System.currentTimeMillis() - startTime
            ParsedWorkoutResult(
                workout = null,
                rawText = rawText,
                status = ParsingStatus.FAILURE,
                modelUsed = "proxy-orchestrator",
                processingTimeMs = duration,
                error = e.message ?: "Unknown proxy error"
            )
        }
    }

    override suspend fun generateSuggestion(context: String): Result<AnalysisSuggestion> =
        runCatching {
            throw UnsupportedOperationException("Proxy analysis not yet supported by backend schema")
        }

    override suspend fun generateSummary(
        period: SummaryPeriod,
        workoutData: String
    ): Result<AnalysisSummary> = runCatching {
        throw UnsupportedOperationException("Proxy analysis not yet supported by backend schema")
    }

    override suspend fun generateExerciseInsight(
        exerciseName: String,
        historyData: String
    ): Result<ExerciseInsight> = runCatching {
        throw UnsupportedOperationException("Proxy analysis not yet supported by backend schema")
    }
}

File: apps/ios/ScribbleFit/ScribbleFit/Features/AI/Data/Engines/ScribbleFitProxyEngine.swift
import Foundation

public final class ScribbleFitProxyEngine: LLMEngine, AnalysisEngine {
    private let networkClient: ScribbleFitNetworkClient
    private let secureKeyStorage: SecureKeyStorage
    private let configRepository: ConfigRepository
    
    public init(
        networkClient: ScribbleFitNetworkClient,
        secureKeyStorage: SecureKeyStorage,
        configRepository: ConfigRepository
    ) {
        self.networkClient = networkClient
        self.secureKeyStorage = secureKeyStorage
        self.configRepository = configRepository
    }
    
    public func parseWorkout(rawText: String) async -> ParsedWorkoutResult {
        let startTime = Date()
        do {
            let token = try await secureKeyStorage.getAuthToken()
            let config = await configRepository.getConfig()
            let systemPrompt = config?.promptText ?? ScribbleFitProxyEngine.defaultPrompt
            
            let request = ParseRequest(rawText: rawText, prompt: systemPrompt)
            let workout = try await networkClient.parseProxy(request: request, token: token)
            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            
            return ParsedWorkoutResult(
                workout: workout,
                rawText: rawText,
                status: .success,
                modelUsed: "proxy-orchestrator",
                processingTimeMs: duration
            )
        } catch {
            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            return ParsedWorkoutResult(
                workout: nil,
                rawText: rawText,
                status: .failure,
                modelUsed: "proxy-orchestrator",
                processingTimeMs: duration,
                error: error.localizedDescription
            )
        }
    }
    
    public func generateSuggestion(context: String) async throws -> AnalysisSuggestion {
        throw NSError(domain: "ScribbleFitProxyEngine", code: 501, userInfo: [NSLocalizedDescriptionKey: "Proxy analysis not yet supported by backend"])
    }
    
    public func generateSummary(period: SummaryPeriod, workoutData: String) async throws -> AnalysisSummary {
        throw NSError(domain: "ScribbleFitProxyEngine", code: 501, userInfo: [NSLocalizedDescriptionKey: "Proxy analysis not yet supported by backend"])
    }
    
    public func generateExerciseInsight(exerciseName: String, historyData: String) async throws -> ExerciseInsight {
        throw NSError(domain: "ScribbleFitProxyEngine", code: 501, userInfo: [NSLocalizedDescriptionKey: "Proxy analysis not yet supported by backend"])
    }
    
    public static let defaultPrompt = """
        You are ScribbleFit AI, a fitness parsing assistant. 
        Your goal is to take raw, messy gym shorthand and parse it into a structured JSON format.
        
        Strictly follow this JSON schema:
        {
          "date": "YYYY-MM-DD",
          "location": "String or null",
          "exercises": [
            {
              "canonical_name": "String",
              "sets": [
                {
                  "weight": number,
                  "reps": integer,
                  "rpe": number or null,
                  "notes": "String or null"
                }
              ]
            }
          ]
        }
        
        Always output valid JSON.
    """
}
